File: src/dp/_494_find_target_sum_ways.rs
struct Solution;

impl Solution {
    pub fn find_target_sum_ways(nums: Vec<i32>, target: i32) -> i32 {
        let v: i32 = nums.iter().sum();
        let v = v.abs();
        if target.abs() > v {
            return 0;
        }
        let mut dp = vec![vec![0; (v * 2 + 1) as usize]; nums.len() + 1];
        dp[0][v as usize] = 1;
        for i in 0..nums.len(){
            for j in 0..dp[i].len() {
                if dp[i][j] != 0 {
                    dp[i+1][j + nums[i] as usize] += dp[i][j];
                    dp[i+1][j - nums[i] as usize] += dp[i][j];
                }
            }
        }
        dp[dp.len() - 1][(v + target) as usize]
    }
}

#[test]
fn test(){
    let v = Solution::find_target_sum_ways(vec![100],-200);
    println!("{}",v);
}

File: hot-100/494_findTargetSumWays.go
package hot_100

import "math"

func findTargetSumWays(nums []int, target int) int {
	sum := 0
	for _, n := range nums {
		sum += n
	}
	if math.Abs(float64(target)) > math.Abs(float64(sum)) {
		return 0
	}
	t := sum*2 + 1
	dp := make([][]int, len(nums))
	for i, _ := range dp {
		dp[i] = make([]int, t)
	}
	if nums[0] == 0 {
		dp[0][sum] = 2
	} else {
		dp[0][sum+nums[0]] = 1
		dp[0][sum-nums[0]] = 1
	}

	for i := 1; i < len(nums); i++ {
		for j := 0; j < t; j++ {
			l, r := 0, 0
			if j-nums[i] >= 0 {
				l = j - nums[i]
			}
			if j+nums[i] < target {
				r = j + nums[i]
			}
			dp[i][j] = dp[i-1][l] + dp[i-1][r]
		}
	}
	return dp[len(nums)-1][sum+target]
}
